benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py

```diff
@@ -0,0 +1,149 @@
+import torch
+import torch.distributed as c10d
+import torch.nn as nn
+from servers.AverageParameterServer import AverageParameterServer
+from torch.nn.parallel import DistributedDataParallel as DDP
+
+from .DdpTrainerBase import DdpTrainerBase
+from .RpcTrainerBase import RpcTrainerBase
+
+
+class DdpSparseRpcTrainer(DdpTrainerBase, RpcTrainerBase):
+
+    PS_MAP = {
+        "AverageParameterServer": AverageParameterServer
+    }
+
+    class HookState:
+
+        def __init__(self, cref, process_group):
+            self.cref = cref
+            self.process_group = process_group
+            self.process_group_size = process_group.size()
+            self.param_loc = 0
+            self.batch_number = -1
+
+        def get_key(self):
+            return f"{self.batch_number},{self.param_loc}"
+
+        def next_batch_state(self):
+            self.param_loc = 0
+            self.batch_number += 1
+
+    def __init__(self, rank, trainer_count, ps_rref, ps_name, backend, use_cuda_rpc, epochs):
+        super().__init__(rank)
+        self.rank = rank
+        self.trainer_count = trainer_count
+        self.ps_rref = ps_rref
+        self.ps = self.PS_MAP[ps_name]
+        self.backend = backend
+        self.use_cuda_rpc = use_cuda_rpc
+        self.epochs = epochs
+
+    @staticmethod
+    def get_tensor_fut(bucket):
+        fut = torch.futures.Future()
+        fut.set_result([bucket.get_tensor()])
+        return fut
+
+    @staticmethod
+    def process_bucket(state, bucket):
+        cref = state.cref
+        tensor = bucket.get_tensor()
+        tensors_count = len(cref.bucket_to_parameters(bucket))
```

Comment:
do we expect only sparse tensors here or can there be dense tensors as well? If sparse-only, there is always just one tensor in the bucket.

Reply:
the process_bucket is designed to process a bucket so it can have sparse or dense tensors.